src/core/reaction_methods/ReactionAlgorithm.hpp

```diff
@@ -109,11 +111,11 @@ class ReactionAlgorithm {
   virtual void on_mc_reject(int &old_state_index){};
   virtual int on_mc_use_WL_get_new_state() { return -10; }
 
-  void make_reaction_attempt(
-      SingleReaction const &current_reaction,
-      std::vector<StoredParticleProperty> &changed_particles_properties,
-      std::vector<int> &p_ids_created_particles,
-      std::vector<StoredParticleProperty> &hidden_particles_properties);
+  std::tuple<std::vector<StoredParticleProperty>, std::vector<int>,
```

Comment:
@jngrad , I found an interesting comment on whether to use custom structs or tuples https://google.github.io/styleguide/cppguide.html#Structs_vs._Tuples linked in https://abseil.io/tips/176
Would it have been better to go for the solution implemented here: https://github.com/espressomd/espresso/pull/4265/commits/7b5bb27d8bc20cb25b43891dc70061a1e2849d65 ?

Reply:
The drawback of `auto` as a return type of a non-inline method is that the return type is only known in the translation unit where the method is defined. If you try to call the method from a child class, you will get
```c++
struct A {
  auto foo();
};
struct B: public A {
  auto bar() { return foo(); }
};
```
```
$> g++ -std=c++14 -c src.hpp
src.hpp: In member function ‘auto B::bar()’:
src.hpp:5:27: error: use of ‘auto A::foo()’ before deduction of ‘auto’
   auto bar() { return foo(); };
```
This means you will eventually need to bring the struct declaration from the method scope to the class scope, when the time comes you want to call the method from outside the translation unit where it is defined, for example when writing a unit test.

I'm not totally convinced wrapping the return values in a struct to provide a namespace is helpful. It also locks us with the names in the struct: if we later decide to rename any of the struct members, we have to update all call sites.

Personally, I find structured bindings sufficiently expressive:
```c++
#include<tuple>
#include<vector>
#include<iostream>

std::tuple<std::vector<int>, std::vector<float>, std::vector<double>> foo();

int main() {
  auto [p_ids_created, hidden_particles, changed_particles] = foo(); // C++17 feature
  std::cout << p_ids_created[0] << " "
            << hidden_particles[0] << " "
            << changed_particles[0] << "\n";
}
```
In this PR the `std::tie(p_ids_created, hidden_particles, changed_particles) = foo()` construction had to be used, because we're stuck with C++14 for hardware reasons. This has the unfortunate consequence that we need to spell out the types of the bound variables before the call to `std::tie()`.

In another place in the PR, you see the following construction:

```c++
#include<array>
#include<vector>
#include<utility>

using Vector3d = std::array<double, 3>;

void place_particle(int pid, Vector3d const &new_pos);

void generate_new_particle_positions(int n_part) {
  using ParticleProperties = std::pair<int, Vector3d>;
  std::vector<ParticleProperties> cache;
  cache.reserve(n_part);
  for (int i = 0; i < n_part; ++pid) {
    auto const pid = i_random(/*...*/);
    auto const new_pos = /*...*/;
    cache.emplace_back(ParticleProperties{pid, new_pos});
  }
  /* ... */
  // restore original particle positions
  for (auto const & item : cache)
    place_particle(std::get<0>(item), std::get<1>(item));
}
```
Of course, I wanted to forward the pair directly to the `place_particle()` function, but it didn't compile because of C++14:
```c++
  // restore original particle positions
  for (auto const &item : cache)
    std::apply(place_particle, item); // C++17 feature
```

Reply:
Thanks for your thoughts! I will think about it a bit more :)